src/ripple/shamap/impl/SHAMap.cpp

```diff
@@ -619,6 +658,43 @@ SHAMap::upper_bound(uint256 const& id) const
     }
     return end();
 }
+SHAMap::const_iterator
+SHAMap::lower_bound(uint256 const& id) const
+{
+    SharedPtrNodeStack stack;
+    walkTowardsKey(id, &stack);
+    while (!stack.empty())
+    {
+        auto [node, nodeID] = stack.top();
+        if (node->isLeaf())
+        {
+            auto leaf = static_cast<SHAMapLeafNode*>(node.get());
+            if (leaf->peekItem()->key() < id)
+                return const_iterator(
+                    this, leaf->peekItem().get(), std::move(stack));
+        }
+        else
+        {
+            auto inner = std::static_pointer_cast<SHAMapInnerNode>(node);
+            for (int branch = selectBranch(nodeID, id) - 1; branch >= 0;
+                 --branch)
+            {
+                if (!inner->isEmptyBranch(branch))
+                {
+                    node = descendThrow(inner, branch);
+                    auto leaf = lastBelow(node, stack, branch);
+                    if (!leaf)
+                        Throw<SHAMapMissingNode>(type_, id);
+                    return const_iterator(
+                        this, leaf->peekItem().get(), std::move(stack));
+                }
+            }
+        }
+        stack.pop();
+    }
+    // TODO: what to return here?
+    return end();
+}
```

Comment:
This function is almost identical to `upper_bound`. We could try to refactor it into a single function. But like I said in the other comment, I don't object to two functions here either.